File: test/language/array_types/python/PackedFixedArrayUInt8Test.py
import os
import zserio

import ArrayTypes

from testutils import getApiDir


class PackedFixedArrayUInt8Test(ArrayTypes.TestCase):
    def testBitSizeOf(self):
        uint8Array = self._createAutoArray()
        packedFixedArray = self.api.PackedFixedArray(uint8Array)
        bitPosition = 2
        fixedArrayBitSize = PackedFixedArrayUInt8Test._calcPackedFixedArrayBitSize()
        self.assertEqual(fixedArrayBitSize, packedFixedArray.bitsizeof(bitPosition))

    def testInitializeOffsets(self):
        uint8Array = self._createAutoArray()
        packedFixedArray = self.api.PackedFixedArray(uint8Array)
        bitPosition = 2
        expectedEndBitPosition = bitPosition + PackedFixedArrayUInt8Test._calcPackedFixedArrayBitSize()
        self.assertEqual(expectedEndBitPosition, packedFixedArray.initialize_offsets(bitPosition))

    def testRead(self):
        writer = zserio.BitStreamWriter()
        self._writePackedFixedArrayToStream(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        packedFixedArray = self.api.PackedFixedArray.from_reader(reader)

        uint8Array = packedFixedArray.uint8_array
        self._checkPackedFixedArray(uint8Array)

    def testWriteRead(self):
        uint8Array = self._createAutoArray()
        packedFixedArray = self.api.PackedFixedArray(uint8Array)
        bitBuffer = zserio.serialize(packedFixedArray)

        self.assertEqual(packedFixedArray.bitsizeof(), bitBuffer.bitsize)
        self.assertEqual(packedFixedArray.initialize_offsets(), bitBuffer.bitsize)

        readPackedFixedArray = zserio.deserialize(self.api.PackedFixedArray, bitBuffer)
        readUint8Array = readPackedFixedArray.uint8_array
        self._checkPackedFixedArray(readUint8Array)

    def testWriteReadFile(self):
        uint8Array = self._createAutoArray()
        packedFixedArray = self.api.PackedFixedArray(uint8Array)
        zserio.serialize_to_file(packedFixedArray, self.BLOB_NAME)

        readPackedFixedArray = zserio.deserialize_from_file(self.api.PackedFixedArray, self.BLOB_NAME)
        readUint8Array = readPackedFixedArray.uint8_array
        self._checkPackedFixedArray(readUint8Array)

    def testWriteWrongArray(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH + 1))
        packedFixedArray = self.api.PackedFixedArray(uint8_array_=uint8Array)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            packedFixedArray.write(writer)

    def _createAutoArray(self):
        return [self.PACKED_ARRAY_ELEMENT] * self.FIXED_ARRAY_LENGTH

    def _checkPackedFixedArray(self, uint8Array):
        self.assertEqual(self.FIXED_ARRAY_LENGTH, len(uint8Array))
        for i in range(self.FIXED_ARRAY_LENGTH):
            self.assertEqual(self.PACKED_ARRAY_ELEMENT, uint8Array[i])

    def _writePackedFixedArrayToStream(self, writer):
        writer.write_bool(True)
        writer.write_bits(self.PACKED_ARRAY_MAX_BIT_NUMBER, 6)
        writer.write_bits(self.PACKED_ARRAY_ELEMENT, 8)

    @staticmethod
    def _calcPackedFixedArrayBitSize():
        bitSize = 1  # packing descriptor: is_packed
        bitSize += 6  # packing descriptor: max_bit_number
        bitSize += 8  # first element

        return bitSize

    BLOB_NAME = os.path.join(getApiDir(os.path.dirname(__file__)), "packed_fixed_array_uint8.blob")

    FIXED_ARRAY_LENGTH = 5
    PACKED_ARRAY_MAX_BIT_NUMBER = 0
    PACKED_ARRAY_ELEMENT = 100

File: test/language/array_types/cpp/PackedFixedArrayUInt8Test.cpp
#include "array_types/packed_fixed_array_uint8/PackedFixedArray.h"
#include "gtest/gtest.h"
#include "zserio/RebindAlloc.h"
#include "zserio/SerializeUtil.h"

namespace array_types
{
namespace packed_fixed_array_uint8
{

using allocator_type = PackedFixedArray::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class PackedFixedArrayUInt8Test : public ::testing::Test
{
protected:
    void fillPackedFixedArray(PackedFixedArray& packedFixedArray)
    {
        auto& uint8Array = packedFixedArray.getUint8Array();
        uint8Array.reserve(FIXED_ARRAY_LENGTH);
        for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
            uint8Array.push_back(PACKED_ARRAY_ELEMENT);
    }

    size_t getPackedFixedArrayBitSize()
    {
        size_t bitSize = 1; // packing descriptor: isPacked
        bitSize += 6; // packing descriptor: maxBitNumber
        bitSize += 8; // firstElement

        return bitSize;
    }

    void checkPackedFixedArray(const PackedFixedArray& packedFixedArray)
    {
        const auto& uint8Array = packedFixedArray.getUint8Array();
        for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        {
            ASSERT_EQ(PACKED_ARRAY_ELEMENT, uint8Array.at(i));
        }
    }

    void writePackedFixedArrayToByteArray(zserio::BitStreamWriter& writer)
    {
        writer.writeBool(true);
        writer.writeBits(PACKED_ARRAY_MAX_BIT_NUMBER, 6);
        writer.writeBits(PACKED_ARRAY_ELEMENT, 8);
    }

    static const std::string BLOB_NAME;

    static const size_t FIXED_ARRAY_LENGTH;
    static const uint8_t PACKED_ARRAY_MAX_BIT_NUMBER;
    static const uint8_t PACKED_ARRAY_ELEMENT;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const std::string PackedFixedArrayUInt8Test::BLOB_NAME = "language/array_types/packed_fixed_array_uint8.blob";

const size_t PackedFixedArrayUInt8Test::FIXED_ARRAY_LENGTH = 5;
const uint8_t PackedFixedArrayUInt8Test::PACKED_ARRAY_MAX_BIT_NUMBER = 0;
const uint8_t PackedFixedArrayUInt8Test::PACKED_ARRAY_ELEMENT = 100;

TEST_F(PackedFixedArrayUInt8Test, bitSizeOf)
{
    PackedFixedArray packedFixedArray;
    fillPackedFixedArray(packedFixedArray);

    const size_t bitPosition = 2;
    const size_t fixedArrayBitSize = getPackedFixedArrayBitSize();
    ASSERT_EQ(fixedArrayBitSize, packedFixedArray.bitSizeOf(bitPosition));
}

TEST_F(PackedFixedArrayUInt8Test, initializeOffsets)
{
    PackedFixedArray packedFixedArray;
    fillPackedFixedArray(packedFixedArray);

    const size_t bitPosition = 2;
    const size_t expectedEndBitPosition = bitPosition + getPackedFixedArrayBitSize();
    ASSERT_EQ(expectedEndBitPosition, packedFixedArray.initializeOffsets(bitPosition));
}

TEST_F(PackedFixedArrayUInt8Test, readConstructor)
{
    zserio::BitStreamWriter writer(bitBuffer);
    writePackedFixedArrayToByteArray(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    PackedFixedArray packedFixedArray(reader);
    checkPackedFixedArray(packedFixedArray);
}

TEST_F(PackedFixedArrayUInt8Test, writeRead)
{
    PackedFixedArray packedFixedArray;
    fillPackedFixedArray(packedFixedArray);

    zserio::BitStreamWriter writer(bitBuffer);
    packedFixedArray.write(writer);

    ASSERT_EQ(packedFixedArray.bitSizeOf(), writer.getBitPosition());
    ASSERT_EQ(packedFixedArray.initializeOffsets(), writer.getBitPosition());

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    PackedFixedArray readPackedFixedArray(reader);
    checkPackedFixedArray(readPackedFixedArray);
}

TEST_F(PackedFixedArrayUInt8Test, writeReadFile)
{
    PackedFixedArray packedFixedArray;
    fillPackedFixedArray(packedFixedArray);

    zserio::serializeToFile(packedFixedArray, BLOB_NAME);

    PackedFixedArray readPackedFixedArray = zserio::deserializeFromFile<PackedFixedArray>(BLOB_NAME);
    checkPackedFixedArray(readPackedFixedArray);
}

TEST_F(PackedFixedArrayUInt8Test, writeWrongArray)
{
    vector_type<uint8_t> uint8Array;
    const size_t wrongArrayLength = FIXED_ARRAY_LENGTH + 1;
    uint8Array.reserve(wrongArrayLength);
    for (size_t i = 0; i < wrongArrayLength; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    PackedFixedArray packedFixedArray;
    packedFixedArray.setUint8Array(uint8Array);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(packedFixedArray.write(writer), zserio::CppRuntimeException);
}

} // namespace packed_fixed_array_uint8
} // namespace array_types
